### ISSUE TITLE
Move internal Container configuration from XML to PHP 

### ISSUE THREAD

**Issue by fabpot** [symfony/symfony#37186]
In Symfony 6, I we will promote usage of configuration written in PHP instead of YAML. For third-party bundles and core, we should do the same, replacing XML with PHP. Doing so would remove the need for the XML lib for core.

The biggest advantage is auto-completion with any modern IDE without explicit support for Symfony, and probably one less thing to learn (how to configure things in YAML/XML).

Be warned that semantic configuration will stay in YAML for now (as using PHP is much harder there).

#36778 did the work for the Twig bundle as an example.

We now need help from the community to move all the other bundles:

**PLEASE, comment here BEFORE starting to work on something and WAIT for the confirmation to avoid duplicate work.**

- [x] src/Symfony/Bundle/DebugBundle/Resources/config/services.xml (@jschaedl)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/annotations.xml (@magnetik)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml (@NguyenTruongLinh)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/cache*.xml (@iamvar)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/collectors.xml (@hvt)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml (@AhmedRaafat14)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/debug*.xml (@TuxBoy)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml (@benji07)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/esi.xml (@tuanminhgp)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/form*.xml (@misekai)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/fragment*.xml (@idetox)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/http*.xml (@IonBazan)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/identity_translator.xml (@smmd)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/lock.xml (@tomasjav)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer*.xml (@instabledesign)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger*.xml (@hyoa)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/mime_type.xml (@GromNaN)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier*.xml (@szepczynski)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/profiling.xml (@hvt)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/property_access.xml (@qneyrat)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/property_info.xml (@qneyrat)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/request.xml (@dangkhoagms)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/routing*.xml (@phamuyentri)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/secrets.xml (@GromNaN)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/security_csrf.xml (@Noweh)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/serializer.xml (@mamontovdmitriy)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml (@rvanlaak)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml (@cocorambo)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/ssi.xml (@50bhan)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/test.xml (@GaryPEGEOT)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/translation*.xml (@malteschlueter)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/validator*.xml (@simivar)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/web*.xml (@ck-developer)
- [x] src/Symfony/Bundle/FrameworkBundle/Resources/config/workflow.xml (@lyrixx)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml (@JudicaelR)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/console.xml (@JudicaelR)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/guard.xml (@JudicaelR)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/security*.xml (@qneyrat)
- [x] src/Symfony/Bundle/SecurityBundle/Resources/config/templating_twig.xml (@ck-developer)
- [x] src/Symfony/Bundle/WebProfilerBundle/Resources/config/*.xml (@jschaedl)

**Comment by szepczynski** [comment]
I do something in Notifier Component few weeks ago. I think  I can migrate  src/Symfony/Bundle/FrameworkBundle/Resources/config/notifier.xml

**Comment by fabpot** [comment]
@szepczynski Great, can you do notifier_transports.xml as well? I've added you name on both :)

**Comment by szepczynski** [comment]
yes, i missed them on the list

**Comment by szepczynski** [comment]
One question about migrating FrameworkExtension - I guess that we need temporary two loaders, one for not migrated XMLs, and one for migrated until all configuration will be migrated.

Can someone add at first Php Loader to FrameworkExtension that everyone who starting from current master don't need create Php Loader?

**Comment by IonBazan** [comment]
I can start with `http_client.xml` and `http_client_debug.xml` first 😅

**Comment by fabpot** [comment]
@szepczynski Can you create the PHP loader in a separate PR that I can merge fast?
@IonBazan added, thanks

**Comment by szepczynski** [comment]
@fabpot ok

**Comment by phamuyentri** [comment]
I want to give it a hand. Can I start with
src/Symfony/Bundle/FrameworkBundle/Resources/config/routing/errors.xml

**Comment by fabpot** [comment]
@phamuyentri Sure, can you do routing*.xml?

**Comment by magnetik** [comment]
Hi,
I can do annotations.xml to help :)

**Comment by fabpot** [comment]
@magnetik Thank you, it's yours now.

**Comment by tomasjav** [comment]
Hi, I'd like to help with lock component.

**Comment by misekai** [comment]
I can help with ` src/Symfony/Bundle/FrameworkBundle/Resources/config/form*.xml`
My understanding is they are 3 files, correct me if I'm wrong:
- ` src/Symfony/Bundle/FrameworkBundle/Resources/config/form.xml`
- ` src/Symfony/Bundle/FrameworkBundle/Resources/config/form_csrf.xml`
- ` src/Symfony/Bundle/FrameworkBundle/Resources/config/form_debug.xml`

**Comment by fabpot** [comment]
@misekai Great! Thank you.

**Comment by fabpot** [comment]
@tomasjav Lock is yours, thank you

**Comment by dangkhoagms** [comment]
I can help with `Bundle/FrameworkBundle/Resources/config/request.xml`

**Comment by AhmedRaafat14** [comment]
@fabpot I can help with `src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml`

**Comment by fabpot** [comment]
@dangkhoagms Thank you
@AhmedRaafat14 Thank you

**Comment by NguyenTruongLinh** [comment]
Hi @fabpot, I think I can help `src/Symfony/Bundle/FrameworkBundle/Resources/config/assets.xml`

**Comment by jschaedl** [comment]
@fabpot I'd like to help with 

* `src/Symfony/Bundle/DebugBundle/Resources/config/services.xml` and
* `src/Symfony/Bundle/WebProfilerBundle/Resources/config/*.xml`

**Comment by fabpot** [comment]
@NguyenTruongLinh assets is yours
@jschaedl wow, that's a lot, thank you.

**Comment by hvt** [comment]
Hey @fabpot, can I do `src/Symfony/Bundle/FrameworkBundle/Resources/config/profiling.xml`?

Was wondering about the namespace of the PHP DI file, why is that different of the file's path?

**Comment by fabpot** [comment]
@hvt ok for profiling.
The namespace allows to use the functions directly without use statements. It's more convenient that way.

**Comment by JudicaelR** [comment]
Hi @fabpot 

Can I start with :
`src/Symfony/Bundle/SecurityBundle/Resources/config/console.xml`
`src/Symfony/Bundle/SecurityBundle/Resources/config/guard.xml`
`src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml`

**Comment by qneyrat** [comment]
@fabpot , i can help with :

- src/Symfony/Bundle/FrameworkBundle/Resources/config/property_access.xml
- src/Symfony/Bundle/FrameworkBundle/Resources/config/property_info.xml

**Comment by fabpot** [comment]
@JudicaelR Cool, thanks for your help
@qneyrat Thank you!

**Comment by tuanminh1997** [comment]
Hi @fabpot ,I can help with:
 src/Symfony/Bundle/FrameworkBundle/Resources/config/property_access.xml

**Comment by fabpot** [comment]
@tuanminh1997 Too late :) Can you choose another one?

**Comment by hyoa** [comment]
@fabpot I can help with  src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger*.xml

**Comment by GaryPEGEOT** [comment]
Hi @fabpot, can I do `src/Symfony/Bundle/FrameworkBundle/Resources/config/test.xml` ?

**Comment by instabledesign** [comment]
Hi @fabpot, can i do `src/Symfony/Bundle/FrameworkBundle/Resources/config/mailer*.xml` ?

**Comment by tuanminh1997** [comment]
Hi @fabpot, can i do src/Symfony/Bundle/FrameworkBundle/Resources/config/esi.xml

**Comment by fabpot** [comment]
@hyoa @GaryPEGEOT @instabledesign @tuanminh1997 You got your files! Thanks.

**Comment by GromNaN** [comment]
Hello, I can do `mime_type.xml` and `secrets.xml`.

**Comment by fabpot** [comment]
@GromNaN Done, thank you.

**Comment by Noweh** [comment]
Hello @fabpot, can I do src/Symfony/Bundle/FrameworkBundle/Resources/config/security_csrf.xml

**Comment by fabpot** [comment]
@Noweh sure, that one is for you.

**Comment by 50bhan** [comment]
Hey @fabpot What about  src/Symfony/Bundle/FrameworkBundle/Resources/config/ssi.xml?

**Comment by hvt** [comment]
I can do `src/Symfony/Bundle/FrameworkBundle/Resources/config/collectors.xml` to if that's ok.

**Comment by benji07** [comment]
Hey @fabpot i can do  `src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml` ?

**Comment by fabpot** [comment]
@hvt Sure
@benji07 Yes, done.

**Comment by smmd** [comment]
Hey, I want to help with `src/Symfony/Bundle/FrameworkBundle/Resources/config/identity_translator.xml` 😃

**Comment by fabpot** [comment]
@smmd It's yours, thank you

**Comment by 50bhan** [comment]
Hey @fabpot Can I go for src/Symfony/Bundle/FrameworkBundle/Resources/config/ssi.xml?

**Comment by fabpot** [comment]
@50bhan Sure, go for it.

**Comment by malteschlueter** [comment]
Hi @fabpot, i think i can help with `src/Symfony/Bundle/FrameworkBundle/Resources/config/translation*.xml`

**Comment by klausi** [comment]
Just a warning here that config files that allow arbitrary code execution are a security threat in certain scenarios. Not sure if this will affect drupal for example that allows config imports over a web interface. Although Symfony service definitions are not used for those imports I think.

Mixing data and code is a bad idea in general, see all the security problems because of mixing HTML and JavaScript or the PHP serialization format.

**Comment by simivar** [comment]
I want to help with src/Symfony/Bundle/FrameworkBundle/Resources/config/validator*.xml.

**Comment by ck-developer** [comment]
I want to help with src/Symfony/Bundle/FrameworkBundle/Resources/config/translation*.xml

**Comment by TuxBoy** [comment]
I can start with  src/Symfony/Bundle/FrameworkBundle/Resources/config/debug*.xml

**Comment by dbrumann** [comment]
> Just a warning here that config files that allow arbitrary code execution are a security threat in certain scenarios. Not sure if this will affect drupal for example that allows config imports over a web interface. Although Symfony service definitions are not used for those imports I think.
> 
> Mixing data and code is a bad idea in general, see all the security problems because of mixing HTML and JavaScript or the PHP serialization format.

@klausi I am not sure, I get your point. You can still mix config styles and use e.g. yaml for exposed (editable) configs. This only affects Symfony's internal configuration which you probably should not expose or change during runtime anyway. Even without malicious intent you can likely break your whole application with it. Can you elaborate on what a possible attack scenario could look like?

**Comment by cocorambo** [comment]
I want to help with src/Symfony/Bundle/FrameworkBundle/Resources/config/session.xml

**Comment by ck-developer** [comment]
Hi @fabpot I want to help with

- src/Symfony/Bundle/FrameworkBundle/Resources/config/translation*.xml
src/Symfony/Bundle/FrameworkBundle/Resources/config/debug*.xml

**Comment by fabpot** [comment]
@malteschlueter @simivar @cocorambo You're good to go.
@ck-developer web*.xml and templating_twig*.xml are yours.

**Comment by rvanlaak** [comment]
@fabpot here to help! Feel free to give me the one or two that are left.

**Comment by fabpot** [comment]
@rvanlaak If you can do src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml, that would be fantastic.

**Comment by idetox** [comment]
@fabpot I can do fragment if it is still left

**Comment by fabpot** [comment]
@idetox Yes, it's yours.

**Comment by qneyrat** [comment]
@fabpot, i can start security* too

**Comment by fabpot** [comment]
@qneyrat that was the last one, thank you.

**Comment by mega6382** [comment]
@fabpot Is there something available for me?

**Comment by fabpot** [comment]
@mega6382 Thanks for asking. All files are now taken care of now.

**Comment by gmponos** [comment]
Hi!... this is a great move.. 

Just an FYI here... I checked with 1-2 installations, and I hope I am not wrong.. but this little thing.. 

![image](https://user-images.githubusercontent.com/5675248/84406182-db300400-ac11-11ea-9ac1-27e6c658c757.png)

does not work with PHP files. For those that don't know, when you click that icon you are supposed to navigate to the service declaration file.

Although I should report this error to PHPStorm.. maybe this might get some attention here as well and there is someone that can apply a fix (hopefully if it is not only my problem)

**Comment by fabpot** [comment]
@gmponos Not PHPStorm user, but I think this is already fixed :) https://github.com/Haehnchen/idea-php-symfony2-plugin/pull/1487

**Comment by gmponos** [comment]
Great.. I believe that ppl will be driven by this move and start moving to PHP service files.. 

I wonder if they will add these to their code coverage

**Comment by nicolas-grekas** [comment]
Please consider sponsoring @Haehnchen if you like the plugin btw!
https://twitter.com/fabpot/status/1271105586726424576

**Comment by yann-eugone** [comment]
I can provide some help on pretty much anything, if something left ! 😅

**Comment by TomasVotruba** [comment]
Cool move :sunglasses:  I like it ! 

I'd like to try it on few packages. Is there some automated script that takes `*.yml` on the input and dumps the same `*.php` file? Something like http://converter.rosstuck.com/

**Comment by yceruto** [comment]
> Is there some automated script that takes *.yml on the input and dumps the same *.php file?

@TomasVotruba -> https://github.com/symfony/maker-bundle/pull/604 (WIP)

**Comment by gggeek** [comment]
I have quite mixed feelings about this.

On one hand, I am all for not using XML: the php config will be probably more efficient to parse, and there will be one less dependency on running Sf.

On the other hand, I find php config to have extremely poor readability for humans, compared to yaml.

Although having IDE support is good, and the `debug` commands help as well, I very often find myself looking for a service definition via using the editor's native 'search' function to grep through source code and taking a quick look at the resulting snippet. The php snippets that I can see f.e. in https://github.com/symfony/symfony/pull/36778/files atm make my head spin and take a long time to parse. It might be that it is just a habit that I need to form, but still... easy to understand by humans rates quite high on my engineering priorities.

**Comment by pounard** [comment]
> On the other hand, I find php config to have extremely poor readability for humans, compared to yaml.

Moreover, indenting as it is done will be reverted automatically by any cs-fixer. I'm afraid it may be a burden to maintain such code.

**Comment by simivar** [comment]
If anyone needs help migrating his files to PHP or won't have time to do that feel free to contact me. I am happy to help :).

**Comment by Tobion** [comment]
We've completed this. Great community effort! Thanks everyone 👏

**Comment by fabpot** [comment]
Thank you all for your great work.

**Comment by adsazad** [comment]
these kind of changes are making me go away from symfony after years of usage.

**Comment by fabpot** [comment]
@adsazad Care to explain. Here, it's all about internal changes, it does not change anything for developers.

**Comment by adsazad** [comment]
Me and my team are really worried about where symfony is going. our lots of project are in symfony when symfony 4 came out there were many changes releated with file structure and doctrine's yml. which made updating project very uneasy.

**Comment by fabpot** [comment]
@adsazad The file structure has not changed at all since 4.0. I'm not sure what you are referring to.

**Comment by adsazad** [comment]
@fabpot. pardon if i was unclear.
i mean update between 3.4 and 4.4

**Comment by wouterj** [comment]
So that update was 2 years ago... The update from 4.4 to 5.0 involved no file structure changes and this PR is updating only 100% internal files. I don't think there will be any major changes in the file structure in the coming feature (before the 4.0 change, the file structure was the same for 6 years (!).

Please also note that the file structure and config formats are not imposed by Symfony. That's something I love about Symfony, you can do and use whatever you feel like using. It's perfectly fine to use the 3.4 file structure with Symfony 5.0.

**Comment by adsazad** [comment]
@wouterj 
I was not really emphasizing only file structure. Don't take this in wrong way but we have started thinking that symfony has started loosing its flexibility.

**Comment by adsazad** [comment]
i just want to say that updating symfony is really hard job.

**Comment by wouterj** [comment]
I don't think you should have that feeling :). I think from a base, Symfony is just as flexible as in the past. However, I agree that one thing changed: The experience when using the default is now much much better by the introduction of Symfony Flex. This hasn't made the other formats worse though, they are still at the same level as when Symfony 2.0 was released.

And to be honest, I think that makes sense. Symfony allows you to get to your destination. When using the paved road, you have a smooth experience to the destination. You can also use the unpaved road, you still get to the destination but the experience is not so smooth. In Symfony 2.0, there was no paved road, in Symfony 4.0+ there is a paved road.

Anyway, yes it was a hard upgrade from the unpaved to the paved road in Symfony 4.0. I think the core team also realizes this. See e.g. this tweet from Fabien (responding to some people that expressed the same feelings as you): https://twitter.com/fabpot/status/1270981253962043392 "I don't envision any revolution in the next few versions either."

In any case, this is completely off-topic for this issue, as this issue is purely internal.

**Comment by TomasVotruba** [comment]
> @adsazad i just want to say that updating symfony is really hard job.

I feel you. I recommend you hire an external upgrade consultant to upgrade it. Doing it manually is hell... and non-sense in 2020. I do around 1 project upgrade a month on average, usually Symfony 3 to 5. With right tools like @rectorphp, coding standard and @phpstan it's piece of cake

**Comment by adsazad** [comment]
this month i did 6 upgrades. i have even developed my own tool for upgrading them.

**Comment by pounard** [comment]
I just wanted to highlight:

> You can also use the unpaved road, you still get to the destination but the experience is not so smooth. In Symfony 2.0, there was no paved road, in Symfony 4.0+ there is a paved road.

I disagree, I used to bend Symfony in all most terrible ways I could think of (starting at 2.8) and it was fine, until 3.4. 3.4 did some great optimizations, and very nice end-user improvements, but it broke some of the natural flexibility it had before. I think it's the natural cost of specialisation and optimisation, nobody can't get around, not even Symfony.

Don't take it wrong, it's not a complaint, I love Symfony and continue gladly to use it, yet in my experience, trying to walk out the paved road since 3.4 is a hard pain, where it was very easy and nice before. Especially with debug component, which is really invasive, and a few other performance optimisations that made the code much harder to read, and broke some undocumented edge cases.

**Comment by gggeek** [comment]
FWIW (and sorry for the OT) I agree with the sentiment that the "paved road" comes with a cost. Increasing the magic while keeping the flexibility leads naturally to greater code complexity. This is also often the case for speed optimizations. The end result is a framework which makes a lot of tedious operations quick and easy, but is harder to grasp when you want to tweak/change bits of it.
Maybe it's just me getting old and taking time to learn all the novelties (aka grumpy), but I am a great fan of code which is easy to read and fits in my head.

**Comment by aleksblendwerk** [comment]
> Here, it's all about internal changes, it does not change anything for developers.

I think this issue has caused some confusion and some people who didn't read the exact words thought it affects *all* configuration files.

Also see here: https://www.reddit.com/r/PHP/comments/h93fjc/symfony_will_now_promote_php_over_yaml_and_xml/.

**Comment by enumag** [comment]
> Be warned that semantic configuration will stay in YAML for now (as using PHP is much harder there).

@fabian Can you be more specific? Which parts of configuration will stay in YAML?

**Comment by Taluu** [comment]
I think he's referring to semantic config (the one you parametrize through `Configuration` class and its `ArrayNodse`, prototypes, and so on)

**Comment by enumag** [comment]
@Taluu As in bundles will still be configured in YAML?

**Comment by chapterjason** [comment]
@enumag You can cofigure these in `php` you just have to load them yourself.

```php
// src/Kernel.php
class Kernel extends BaseKernel
{
    use MicroKernelTrait;

    protected function prepareContainer(ContainerBuilder $container)
    {
        parent::prepareContainer($container);

        (require dirname(__DIR__) . '/config/packages/workflow.php')($container);
    }
}
```

```php
// config/packages/workflow.php
return static function (ContainerBuilder $container) {
    $container->loadFromExtension(
        'framework',
        [
            'workflows' => [
                ChangeEmailWorkflow::NAME => [
                    "type" => "state_machine",
                    "marking_store" => [
                        "type" => "method",
                        "property" => "state",
                    ],
                    "supports" => [Token::class],
                    "initial_marking" => ChangeEmailWorkflow::STATE_PENDING,
                    "places" => [
                        ChangeEmailWorkflow::STATE_PENDING,
                        ChangeEmailWorkflow::STATE_APPROVED,
                        ChangeEmailWorkflow::STATE_VERIFIED,
                    ],
                    "transitions" => [
                        ChangeEmailWorkflow::TRANSITION_APPROVE => [
                            "from" => ChangeEmailWorkflow::STATE_PENDING,
                            "to" => ChangeEmailWorkflow::STATE_APPROVED,
                        ],
                        ChangeEmailWorkflow::TRANSITION_VERIFY => [
                            "from" => ChangeEmailWorkflow::STATE_APPROVED,
                            "to" => ChangeEmailWorkflow::STATE_VERIFIED,
                        ],
                    ],
                ],
            ],
        ]
    );
};
```
